flutter/lib/src/profiling.dart

```diff
@@ -0,0 +1,115 @@
+import 'dart:async';
+
+// ignore: implementation_imports
+import 'package:sentry/src/profiling.dart';
+// ignore: implementation_imports
+import 'package:sentry/src/sentry_envelope_item_header.dart';
+// ignore: implementation_imports
+import 'package:sentry/src/sentry_item_type.dart';
+
+import '../sentry_flutter.dart';
+import 'native/sentry_native.dart';
+
+// ignore: invalid_use_of_internal_member
+class NativeProfilerFactory implements SentryProfilerFactory {
+  final SentryNative _native;
+  final ClockProvider _clock;
+
+  NativeProfilerFactory(this._native, this._clock);
+
+  static void attachTo(Hub hub, SentryNative native) {
+    // ignore: invalid_use_of_internal_member
+    final options = hub.options;
+
+    // ignore: invalid_use_of_internal_member
+    if ((options.profilesSampleRate ?? 0.0) <= 0.0) {
+      return;
+    }
+
+    if (options.platformChecker.isWeb) {
+      return;
+    }
+
+    if (options.platformChecker.platform.isMacOS ||
+        options.platformChecker.platform.isIOS) {
+      // ignore: invalid_use_of_internal_member
+      hub.profilerFactory = NativeProfilerFactory(native, options.clock);
+    }
+  }
+
+  @override
+  NativeProfiler? startProfiler(SentryTransactionContext context) {
+    if (context.traceId == SentryId.empty()) {
+      return null;
+    }
+
+    final startTime = _native.startProfiler(context.traceId);
+    if (startTime == null) {
+      return null;
+    }
+    return NativeProfiler(_native, startTime, context.traceId, _clock);
```

Comment:
just to be sure, is the `profilesSampleRate` being checked by the native profiler?

Reply:
It's used by the the hub when a transaction is started:

```dart
      SentryProfiler? profiler;
      if (_profilerFactory != null &&
          _tracesSampler.sampleProfiling(samplingDecision)) {
        profiler = _profilerFactory?.startProfiler(transactionContext);
      }
```

```dart
  bool sampleProfiling(SentryTracesSamplingDecision tracesSamplingDecision) {
    double? optionsRate = _options.profilesSampleRate;
    if (optionsRate == null || !tracesSamplingDecision.sampled) {
      return false;
    }
    return _sample(optionsRate);
  }
```